### PR TITLE
[434982] Escape Java Keywords inside dispatch if else cascade

### PR DESCRIPTION
**Pull request by cdietrich**
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=434982

**Comment by cdietrich** [comment]
Jenkins test

**Comment by cdietrich** [comment]
can someone review this change

**Comment by svenefftinge** [comment]
+1

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java [hunk 1]
```diff
@@ -205,11 +209,14 @@ protected void generateActualDispatchCall(JvmOperation dispatchOperation, JvmOpe
...
 	}
 
-	/**
-	 * TODO: we assume that names from the inferred JVM model are the same in the Java code. Does that assumption
-	 * hold?
-	 */
 	protected String getVarName(JvmIdentifiableElement ex, IAppendable appendable) {
-		return ex.getSimpleName();
+		final String name = ex.getSimpleName();
+		if (name == null) {
```
Comment (akosyakov) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: When does it happen? Could we provider a test for this case?

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java [hunk 2]
```diff
@@ -205,11 +209,14 @@ protected void generateActualDispatchCall(JvmOperation dispatchOperation, JvmOpe
...
+			return "__unknown__";
+		} else if (keywords.isJavaKeyword(name)) {
```
Comment (akosyakov) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: Could we replace else-if cascade just with if check?
if (keywords.isJavaKeyword(name)) {
	return name + '_';
}
return name;

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java [hunk 3]
```diff
@@ -205,11 +209,14 @@ protected void generateActualDispatchCall(JvmOperation dispatchOperation, JvmOpe
...
+			return name+"_";
```
Comment (akosyakov) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: What if there is another argument with such name, e.g. 'assert_'? Would be generated code valid?

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 1 above)
Comment (cdietrich) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: see testReservedJavaKeywords

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 3 above)
Comment (cdietrich) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: well we have the same behavior implemented for method parameters

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 1 above)
Comment (akosyakov) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: I meant for the case when generated code contains '__unknown__'.

Comment (cdietrich) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: i just lazily copy & pasted the if else cascade

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 3 above)
Comment (akosyakov) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: I've just tried it.
Input:
class Foo {
	def dispatch foo(Object assert, Object assert_) {}
        def dispatch foo(String assert, Object assert_) {}
}

Output:
public Object foo(final Object assert_, final Object assert__1) {
    if (assert_ instanceof String) {
      return _foo((String)assert_, **assert_**);
    } else if (assert_ != null) {
      return _foo(assert_, **assert_**);
    } else {
      throw new IllegalArgumentException("Unhandled parameter types: " +
        Arrays.<Object>asList(assert_, **assert_**).toString());
    }
  }

It seems that bogus second argument is used inside the dispatch method.

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 1 above)
Comment (akosyakov) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: From where did you copy it? Could you point me?

File: plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java (hunk 3 above)
Comment (akosyakov) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: @cdietrich I reopened the associated bug, you can have a look at it if you want

Comment (cdietrich) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: sry i right busy now. can you have a look. i took it from
org.eclipse.xtext.xbase.compiler.JvmModelGenerator.makeJavaIdentifier(String)

Comment (akosyakov) [inline plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java]: @cdietrich it is not so urgent, you can have a look when you have time

**Pull request merged**